gensim/models/nmf.py

```diff
@@ -548,10 +555,15 @@ def _setup(self, corpus):
 
         if isinstance(corpus, scipy.sparse.csc.csc_matrix):
             first_doc = corpus.getcol(0)
-        else:
-            first_doc_it = itertools.tee(corpus, 1)
-            first_doc = next(first_doc_it[0])
+        elif hasattr(corpus, "__iter__"):
```

Comment:
This is not the right way to test for an iterable in Python. Also, still destructive in case the corpus stream is a non-repeatable generator.

Check out `utils.is_corpus` instead.

Reply:
I test for an iterable here, the next line tests for an iterator.

Thanks, I'll use `utils` method.

Reply:
I test for an iterable here, the next line tests for an iterator.

I need to distinguish between iterator and an ordinary iterable, hence can't use the utils.is_corpus method.

Reply:
Radim, NMF works correctly with repeatable iterables, non-repeatable iterables (iterators) and csc matrices. It does arbitrary number of passes in the first case, only one pass in the second case and arbitrary number of passes in the third case.

Reply:
What do you mean by arbitrary? Corpus passes are expensive, we should know exactly how many passes there are.

Reply:
Pardon my english, I mean that you can make as much passes as you want.

Reply:
I see. What happens when I specify 2 passes but supply a generator input?

Reply:
It will make 2 passes, as was intended.

So, if we can re-iterate over something - we can make multiple passes, if not - it will just do one pass and then stop.

Reply:
I'll mention that behavior in the description of the `pass` parameter

Reply:
I cannot make 2 passes over a generator :)

I think the proper action when user specifies 2 passes, but only supplies a non-restartable sequence, is to raise a `ValueError`.

Reply:
I thought you meant generator function, like if we call `__iter__`, and it creates an iterator every time.

Anyway, I agree, I'll make it so that NMF will raise a `ValueError` in that case.